### ISSUE TITLE
Why `softbreak` and `space` has no location information?

### ISSUE THREAD

**Issue by pddg** [tamuratak/latex-utensils#26]
Since v4.0.0, `softbreak` and `space` have been introduced. However, they do not have location information.
If they indicate a single character, the positional information should be uniquely identifiable, so there seems no need to make them undefined.

https://github.com/tamuratak/latex-utensils/blob/921776d26b94db847ed54fe111bde8592b3b87b5/src/latex/latex_parser_types.ts#L155
https://github.com/tamuratak/latex-utensils/blob/921776d26b94db847ed54fe111bde8592b3b87b5/src/latex/latex_parser_types.ts#L164

**Comment by tamuratak** [comment]
Please explain why you need the `location` property for `softbreak` and `space` with examples.

**Comment by pddg** [comment]
We are working on a tool to convert the AST output of latex-utensils to AST of [textlint](https://github.com/textlint/textlint).
We are required to completely reconstruct the original text from AST of textlint. This includes whitespace, line breaks, and comments.

Consider the following sample.

```tex
abc def
% comment
ghi
```

The AST output by latex-utensils was missing some information such as whitespace, so we needed to fill in some information in order to reconstruct the text completely.
Until v3.0.0, we first inserted the comment nodes at their original locations in the AST, then recomputed the missing parts (`space` and `softbreak` since v4.0.0 or later) from the location of each node and inserted them. This is mostly working well now.

If we want to do the same thing in v4.0.0 or later, there are two possible ways.

1. Eliminate `softbreak` and `space` altogether before processing.
2. Estimate the location of the `softbreak` and `space` and add the location information to it.

### Case 1

In case 1, assuming that the introduction of `softbreak` and `space` does not cause destructive changes in the parsing of other nodes, we only need to increase our processing a little. We already know that they are there, but we will have to discard them, which will be wasteful. :(

### Case 2

In case 2, since location is undefined by definition, the type will change and a dirty workaround may be required.

When processing after converting to a different type (in this case, textlint's AST Node), it is necessary to first embed a dummy location, insert the comment node to AST, and then estimate the location information. This is because when a comment comes immediately after `space`, it is not known how many whitespaces `space` contains until the comment is inserted. Also, it is not yet clear if this will work perfectly.

### Conclusions

The location is not always necessary, but it would be nice to have it so that we can skip some work. I'd appreciate it if you could implement it when you have time 🙏.

**Comment by tamuratak** [comment]
I am not convinced.

Please explain in the following steps.

1. What kind of features do you want to implement? Please explain with sample LaTeX documents.
2. How do you want to implement the feature? Please explain with links to actual codes.
3. Why do you need the location property? Can you implement the feature without it?

**Comment by pddg** [comment]
### 1

We maintain a tool to convert the AST to textlint's AST while adding necessary information so that the original text can be recovered from the AST in its complete form.

```tex
abc def
% comment
ghi
```

The parsing result of it is as follows. To describe the current implementation, we will use the parsing results as of v3.0.0.

<details><summary>npx luparse -lc sample.tex</summary>

```json
{
  "kind": "ast.root",
  "content": [
    {
      "kind": "text.string",
      "content": "abc",
      "location": {
        "start": {
          "offset": 0,
          "line": 1,
          "column": 1
        },
        "end": {
          "offset": 3,
          "line": 1,
          "column": 4
        }
      }
    },
    {
      "kind": "text.string",
      "content": "def",
      "location": {
        "start": {
          "offset": 4,
          "line": 1,
          "column": 5
        },
        "end": {
          "offset": 7,
          "line": 1,
          "column": 8
        }
      }
    },
    {
      "kind": "text.string",
      "content": "ghi",
      "location": {
        "start": {
          "offset": 18,
          "line": 3,
          "column": 1
        },
        "end": {
          "offset": 21,
          "line": 3,
          "column": 4
        }
      }
    }
  ],
  "comment": [
    {
      "kind": "comment",
      "content": " comment",
      "location": {
        "start": {
          "offset": 8,
          "line": 2,
          "column": 1
        },
        "end": {
          "offset": 18,
          "line": 3,
          "column": 1
        }
      }
    }
  ]
}
```

</details>

Step1: all nodes are converted to AST Node of textlint. Since textlint requires location for most nodes, it inherits the location before conversion.

<details><summary>AST of Step1</summary>

```json
{
  "type": "Document",
  "range": [
    0,
    22
  ],
  "loc": {
    "start": {
      "line": 1,
      "column": 0
    },
    "end": {
      "line": 3,
      "column": 3
    }
  },
  "children": [
    {
      "loc": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": 3,
          "column": 3
        }
      },
      "range": [
        0,
        21
      ],
      "type": "Paragraph",
      "children": [
        {
          "loc": {
            "start": {
              "line": 1,
              "column": 0
            },
            "end": {
              "line": 1,
              "column": 3
            }
          },
          "range": [
            0,
            3
          ],
          "type": "Str",
          "value": "abc"
        },
        {
          "loc": {
            "start": {
              "line": 1,
              "column": 4
            },
            "end": {
              "line": 1,
              "column": 7
            }
          },
          "range": [
            4,
            7
          ],
          "type": "Str",
          "value": "def"
        },
        {
          "loc": {
            "start": {
              "line": 3,
              "column": 0
            },
            "end": {
              "line": 3,
              "column": 3
            }
          },
          "range": [
            18,
            21
          ],
          "type": "Str",
          "value": "ghi"
        }
      ]
    }
  ]
}
```

</details>

Step2: Insert the comment node where it belongs. This will allow us to restore the text, including the comments, by tracing back from the root of AST. 
Traverse AST in order to find the position where the comment should be inserted. Then, we can find the position to insert the comment by just tracing it in order.

<details><summary>AST of Step2</summary>

```json
{
  "type": "Document",
  "range": [
    0,
    22
  ],
  "loc": {
    "start": {
      "line": 1,
      "column": 0
    },
    "end": {
      "line": 3,
      "column": 3
    }
  },
  "children": [
    {
      "loc": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": 3,
          "column": 3
        }
      },
      "range": [
        0,
        21
      ],
      "type": "Paragraph",
      "children": [
        {
          "loc": {
            "start": {
              "line": 1,
              "column": 0
            },
            "end": {
              "line": 1,
              "column": 3
            }
          },
          "range": [
            0,
            3
          ],
          "type": "Str",
          "value": "abc"
        },
        {
          "loc": {
            "start": {
              "line": 1,
              "column": 4
            },
            "end": {
              "line": 1,
              "column": 7
            }
          },
          "range": [
            4,
            7
          ],
          "type": "Str",
          "value": "def"
        },
        {
          "loc": {
            "start": {
              "line": 2,
              "column": 0
            },
            "end": {
              "line": 3,
              "column": 0
            }
          },
          "range": [
            8,
            18
          ],
          "value": " comment",
          "type": "Comment"
        },
        {
          "loc": {
            "start": {
              "line": 3,
              "column": 0
            },
            "end": {
              "line": 3,
              "column": 3
            }
          },
          "range": [
            18,
            21
          ],
          "type": "Str",
          "value": "ghi"
        }
      ]
    }
  ]
}
```

</details>

Step3: we need to supplement a few lost parts. Now we have lost information about whitespaces and line breaks, which can be detected from inconsistencies in the position between nodes.
For example, from the location information of `abc` and `def` nodes, we know that there should be a character between them, so we extract the character at that position from the original text and add it.

<details><summary>AST of Step3</summary>

```json
{
  "type": "Document",
  "range": [
    0,
    22
  ],
  "loc": {
    "start": {
      "line": 1,
      "column": 0
    },
    "end": {
      "line": 3,
      "column": 3
    }
  },
  "children": [
    {
      "loc": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": 3,
          "column": 3
        }
      },
      "range": [
        0,
        21
      ],
      "type": "Paragraph",
      "children": [
        {
          "loc": {
            "start": {
              "line": 1,
              "column": 0
            },
            "end": {
              "line": 1,
              "column": 3
            }
          },
          "range": [
            0,
            3
          ],
          "type": "Str",
          "value": "abc"
        },
        {
          "loc": {
            "start": {
              "line": 1,
              "column": 3
            },
            "end": {
              "line": 1,
              "column": 4
            }
          },
          "range": [
            3,
            4
          ],
          "type": "Html",
          "value": " "
        },
        {
          "loc": {
            "start": {
              "line": 1,
              "column": 4
            },
            "end": {
              "line": 1,
              "column": 7
            }
          },
          "range": [
            4,
            7
          ],
          "type": "Str",
          "value": "def"
        },
        {
          "loc": {
            "start": {
              "line": 1,
              "column": 7
            },
            "end": {
              "line": 2,
              "column": 0
            }
          },
          "range": [
            7,
            8
          ],
          "type": "Html",
          "value": "\n"
        },
        {
          "loc": {
            "start": {
              "line": 2,
              "column": 0
            },
            "end": {
              "line": 3,
              "column": 0
            }
          },
          "range": [
            8,
            18
          ],
          "value": " comment",
          "type": "Comment"
        },
        {
          "loc": {
            "start": {
              "line": 3,
              "column": 0
            },
            "end": {
              "line": 3,
              "column": 3
            }
          },
          "range": [
            18,
            21
          ],
          "type": "Str",
          "value": "ghi"
        }
      ]
    }
  ]
}
```

</details>


### 2

The code corresponding to each of the steps described above is as follows

Step1: https://github.com/textlint/textlint-plugin-latex2e/blob/a9853168619a976b809fc17924fcd20edf3c02b6/src/latex-to-ast.ts#L31-L523
Step2: https://github.com/textlint/textlint-plugin-latex2e/blob/a9853168619a976b809fc17924fcd20edf3c02b6/src/completeComment.ts
Step3: https://github.com/textlint/textlint-plugin-latex2e/blob/a9853168619a976b809fc17924fcd20edf3c02b6/src/completeBlank.ts

If `softbreak` and `space` had location information, Step 3 might be unnecessary.

### 3

As I mentioned above, if it has location information, Step 3 will become probably unnecessary. Also, the `location: undefined` Node is gone, which makes it easier to handle.
In other words, it's not essential, but it's nice to have :)

**Comment by tamuratak** [comment]
やはりよくわからないので日本語で説明してください。

**Comment by tamuratak** [comment]
> In other words, it's not essential, but it's nice to have :)

つまり必ずしも必要ではないということですか。

**Comment by pddg** [comment]
伝わりにくくて申し訳ないです🙏 
textlintをLaTeXソースに対しても実行できるようにするため、LaTeXのASTからtextlint用のASTに変換するプラグインを実装しています。textlintではlintだけでなく自動修正にも対応しており、textlintのASTからは元のソースを再構成できる必要があります。そのため文法的に意味のない空白や改行及びコメントもそこに含める必要があります。

> つまり必ずしも必要ではないということですか。

はい。前処理を施して含まれる `softbreak` や` space` を削除してしまえば、v4.0.0以降で出力されるASTもv3.0.0に近い物になるからです。

v3.0.0を利用している現在は、ASTに含まれていないスペースや改行などを、他のノードのlocation情報を突き合わせて存在しない部分を見つけることで補っています。`abc def` をパースすると `abc` のノードと `def` のノードが出てきますが、それらのlocationを見ると間の一文字がないことが分かるため、元の文章のその位置の文字列（つまり空白）を変換したtextlintのASTに組み込むようにしています。改行やその他latex-utensilsがASTで表現しなかったものは、基本的にこの方法でtextlintのASTに変換しています。

v4.0.0以降ではその場所に`softbreak`や`space`が追加されています。しかし、location情報がなくそれだけでは元の文章を完全に復元できるASTにならないため、我々のケースでは利用することが難しいです。
例えば `space` には複数の空白が含まれうるのですが、その情報は失われてしまっており、結局周囲のノードと突き合わせていくつの空白を表現しているのかを突き止める必要があります。前述の方法で補うことはできますが、locationがあればその不要な操作を避けることができます。

**Comment by tamuratak** [comment]
> ASTからは元のソースを再構成できる必要があります。

このような用途はいまのところ想定していないので、提案は reject ということにしたいと思います。
